lib/plugins/environments.js

```diff
@@ -272,4 +273,79 @@ module.exports = class Environments extends Diffable {
             environment_name: existing.name
         });
     }
-}
\ No newline at end of file
+
+    sync () {
+      const resArray = []
+      if (this.entries) {
+        let filteredEntries = this.filterEntries()
+        return this.find().then(existingRecords => {
+
+          // Filter out all empty entries (usually from repo override)
+          for (const entry of filteredEntries) {
+            for (const key of Object.keys(entry)) {
+              if (entry[key] === null || entry[key] === undefined) {
+                delete entry[key]
+              }
+            }
+          }
+          filteredEntries = filteredEntries.filter(entry => Object.keys(entry).filter(key => !MergeDeep.NAME_FIELDS.includes(key)).length !== 0)
+
+          const changes = []
+
+          existingRecords.forEach(x => {
+            if (!filteredEntries.find(y => this.comparator(x, y))) {
+              const change = this.remove(x).then(res => {
+                if (this.nop) {
+                  return resArray.push(res)
+                }
+                return res
+              })
+              changes.push(change)
+            }
+          })
+
+          filteredEntries.forEach(attrs => {
+            const existing = existingRecords.find(record => {
+              return this.comparator(record, attrs)
+            })
+
+            if (!existing) {
+              const change = this.add(attrs).then(res => {
+                if (this.nop) {
+                  return resArray.push(res)
+                }
+                return res
+              })
+              changes.push(change)
+            } else if (this.changed(existing, attrs)) {
+              const change = this.update(existing, attrs).then(res => {
+                if (this.nop) {
+                  return resArray.push(res)
+                }
+                return res
+              })
+              changes.push(change)
+            }
+          })
+
+          if (this.nop) {
+            return Promise.resolve(resArray)
+          }
+          return Promise.all(changes)
+        }).catch(e => {
+          if (this.nop) {
+            if (e.status === 404) {
+              // Ignore 404s which can happen in dry-run as the repo may not exist.
+              return Promise.resolve(resArray)
+            } else {
+              resArray.push(new NopCommand(this.constructor.name, this.repo, null, `error ${e} in ${this.constructor.name} for repo: ${JSON.stringify(this.repo)} entries ${JSON.stringify(this.entries)}`, 'ERROR'))
```

Comment:
'NopCommand' is not defined here

Reply:
Thanks @klutchell .  I have added a pair of commits to address this: [cf3fa72](https://github.com/github/safe-settings/pull/646/commits/cf3fa7208e4123c59eeb379e18d4d179fca10ca3) and [caa7d04](https://github.com/github/safe-settings/pull/646/commits/caa7d04041f78632575a36abfee4f4e63c9ba841)
With these additions, the NopCommand is now defined.  Additionally, there was an argument missing from the constructor for the environments plugin when instantiated from the unit tests.  That has also been corrected.

I appreciate that you're taking a look at the PR.  I have more to follow on from this one.

Reply:
Any improvements to the environments plugin at this point is appreciated, as we are struggling to use this project to set environment level variables.

With PR https://github.com/github/safe-settings/pull/616 we were able to get it to create the initial set of environments and the respective variables, but now we can't get it to detect changes to those variables in the configuration.

Reply:
I'm not the project maintainer, but that sounds like the correct approach to me.

WDYT @decyjphr ?